### PR TITLE
F#25 enemy ai controller interface

### PR DESCRIPTION
**Pull request by danielbe1**
* Implementation is based on behavior trees from [libGDX AI](https://github.com/libgdx/gdx-ai/wiki/Behavior-Trees#core-concepts)
* Core Concepts:
    * Actor - Any game entity that has behavior (NPCs, possibly even players)
    * Actor Controller - Class responsible for adapting the actor to its current state
    * Action - Represents a change in the state of an actor, e.g. Walk, Jump, Attack, will usually also cause a change to animation, velocity etc
    * Task - Represent concrete goals for an actor to achieve through the application of actions over time
    * Behavior Tree - Responsible for selecting tasks for actors

### CODE REVIEW

File: Assets/Scripts/Actors/Actions/ActorAnimationAction.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    private readonly Vector2 _velocity;
+    
+    protected ActorAnimationAction(Vector2 velocity)
+    {
+        _velocity = velocity;
+    }
+
+    public bool Execute(IActor actor)
+    {
+        MaybeGetAnimation(actor).IfSome(actor.SetAnimation);
```
Comment (massieb) [inline Assets/Scripts/Actors/Actions/ActorAnimationAction.cs]: This reads very hard. What happens here. Does the extending class decides if there is an animation and is the animation then set?

File: Assets/Scripts/Actors/Actions/ActorAnimationAction.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        actor.SetVelocity(_velocity);
+
+        return true;
+    }
+
+    protected abstract Option<ActorAnimation> MaybeGetAnimation(IActor actor);
```
Comment (massieb) [inline Assets/Scripts/Actors/Actions/ActorAnimationAction.cs]: Maybe answer my question by adding comments here what you expect to get from this method

File: Assets/Scripts/Actors/Ai/Guards/AboutToDropGuard.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+using Ai.BehaviorTree.Nodes;
+using UnityEngine;
+
+public class AboutToDropGuard : BtGuard<IActor>
+{
+    public bool Process(IActor data)
```
Comment (massieb) [inline Assets/Scripts/Actors/Ai/Guards/AboutToDropGuard.cs]: rename data to actor

Comment (massieb) [inline Assets/Scripts/Actors/Ai/Guards/AboutToDropGuard.cs]: Method name does not indicate that something is returned

File: Assets/Scripts/Actors/Ai/Tasks/HorizontalPatrolTask.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+using UnityEngine;
+
+/// <summary>
+/// A task that continuously causes an actor to move horizontally between two points.
+/// </summary>
+public class HorizontalPatrolTask : BtTaskNode<IActor>
+{
+    private readonly ActorMoveHorizontallyAction _runRight, _runLeft;
+
+    private readonly float _x1, _x2;
```
Comment (massieb) [inline Assets/Scripts/Actors/Ai/Tasks/HorizontalPatrolTask.cs]: rename to position1, position2 or leftPosition, rightPosition?

File: Assets/Scripts/Actors/Ai/Tasks/HorizontalPatrolTask.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                break;
+            case State.MoveToX2:
+                UpdateMovementTowardsPoint(_x2, position.x, data, State.MoveToX1);
+                break;
+        }
+
+        return BtResult.Running;
+    }
+
+    protected override BtResult Initialize(IActor data)
```
Comment (massieb) [inline Assets/Scripts/Actors/Ai/Tasks/HorizontalPatrolTask.cs]: rename data to actor

File: Assets/Scripts/Actors/Ai/Tasks/HorizontalPatrolTask.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            StartMovementTowardsPoint(_x2, position.x, data);
+            _state = State.MoveToX2;
+        }
+        
+        return BtResult.Success;
+    }
+
+    protected override void CleanAfterSuccess(IActor data)
+    {
```
Comment (massieb) [inline Assets/Scripts/Actors/Ai/Tasks/HorizontalPatrolTask.cs]: throw exceptions for not implemented methods? Or are these methods used by the library code? Then comment that this method is left empty for a reason.

File: Assets/Scripts/Actors/CollisionSensor.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    [RequireComponent(typeof(Collider2D))]
+    public class CollisionSensor : MonoBehaviour
+    {
+        [ReadOnly]
+        public int TriggeringLayer;
+
+        [ShowNonSerializedField]
+        private int triggersTouched;
+
+        public bool Triggered => triggersTouched > 0;
```
Comment (massieb) [inline Assets/Scripts/Actors/CollisionSensor.cs]: does this update? or only initializes

File: Assets/Scripts/Actors/IActor.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    ActorCapabilities GetCapabilities();
+
+    void SetAnimation(ActorAnimation animation);
+    
+    bool IsTouchingCeiling();
+    
+    bool IsTouchingGround();
+
+    bool IsAboutToDrop();
+
```
[the remark this answers is not in the document]
Comment (danielbe1) [reply Assets/Scripts/Actors/IActor.cs]: 👍

File: Assets/Scripts/Common/Range.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        else
+        {
+            _min = max;
+            _max = min;
+        }
+    }
+
+    public bool Equals(RangeFloat other)
+    {
+        return _min == other._min && _max == other._max;
```
[the remark this answers is not in the document]
Comment (danielbe1) [reply Assets/Scripts/Common/Range.cs]: 👍

File: Assets/Scripts/Actors/CollisionSensor.cs (hunk 7 above)
Comment (massieb) [reply Assets/Scripts/Actors/CollisionSensor.cs]: Ohh so its executed everytime you ask for Triggered?
Than i would rename that to IsTriggered

File: Assets/Scripts/Common/Range.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    public override int GetHashCode()
+    {
+        unchecked
+        {
+            return (_min.GetHashCode() * 397) ^ _max.GetHashCode();
+        }
+    }
+}
+
+#if UNITY_EDITOR
```
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Common/Range.cs]: 👍

File: Assets/Scripts/Firing/Weapon.cs [hunk 11]
```diff
@@ -33,4 +33,17 @@ public void Shoot()
         }
     }
 
+    public void Shoot(Vector2 bulletPosition, HorizontalDirection direction)
```
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Firing/Weapon.cs]: 👍

File: Assets/Scripts/Level/ParallaxBackgroundLayer.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
+using UnityEngine;
+
+public class ParallaxBackgroundLayer
```
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Level/ParallaxBackgroundLayer.cs]: 👍

File: Assets/Scripts/Actors/Actions/ActorAnimationAction.cs (hunk 2 above)
Comment (massieb) [reply Assets/Scripts/Actors/Actions/ActorAnimationAction.cs]: 👍

File: Assets/Scripts/Actors/IActor.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Actors/IActor.cs]: 👍

File: Assets/Scripts/Actors/CollisionSensor.cs (hunk 7 above)
Comment (danielbe1) [reply Assets/Scripts/Actors/CollisionSensor.cs]: 👍

**Pull request merged**